File: test/array_test.cpp
#include "../array.h"

#include <gtest/gtest.h>
#include <cstdio>

TEST(ArrayTest, Exceptions) {
    Array arr;
    EXPECT_THROW(arr.get(-1), std::out_of_range);
    EXPECT_THROW(arr.get(0), std::out_of_range);
    EXPECT_THROW(arr.removeAt(0), std::out_of_range);
    EXPECT_THROW(arr.insertAt(1, "x"), std::out_of_range);
}

TEST(ArrayTest, SaveLoadTextBinary) {
    Array arr;
    arr.pushBack("a");
    arr.pushBack("b");
    arr.saveToText("arr.txt");
    arr.saveToBinary("arr.bin");

    Array arr2;
    arr2.loadFromText("arr.txt");
    EXPECT_EQ(arr2.getSize(), 2);
    EXPECT_EQ(arr2.get(0), "a");

    Array arr3;
    arr3.loadFromBinary("arr.bin");
    EXPECT_EQ(arr3.getSize(), 2);
    EXPECT_EQ(arr3.get(1), "b");
    
    // Удаляем тестовые файлы
    remove("arr.txt");
    remove("arr.bin");
}

TEST(ArrayTest, BasicOperations) {
    Array arr;
    arr.pushBack("1");
    arr.pushBack("3");
    arr.pushFront("0");
    arr.insertAt(2, "2");

    EXPECT_EQ(arr.getSize(), 4);
    EXPECT_EQ(arr.get(0), "0");
    EXPECT_EQ(arr.get(1), "1");
    EXPECT_EQ(arr.get(2), "2");
    EXPECT_EQ(arr.get(3), "3");

    arr.popFront();
    arr.popBack();
    EXPECT_EQ(arr.getSize(), 2);
}

TEST(ArrayTest, FindAndSet) {
    Array arr;
    arr.pushBack("a");
    arr.pushBack("b");
    arr.pushBack("c");
    
    EXPECT_EQ(arr.find("b"), 1);
    EXPECT_EQ(arr.find("d"), -1);
    
    arr.set(1, "x");
    EXPECT_EQ(arr.get(1), "x");
}

TEST(ArrayTest, RemoveAtValid) {
    Array arr;
    arr.pushBack("a");
    arr.pushBack("b");
    arr.pushBack("c");
    
    arr.removeAt(1);
    EXPECT_EQ(arr.getSize(), 2);
    EXPECT_EQ(arr.get(0), "a");
    EXPECT_EQ(arr.get(1), "c");
}

TEST(ArrayTest, PopOperationsOnEmpty) {
    Array arr;
    arr.popBack(); // Не должно падать
    arr.popFront(); // Не должно падать
    EXPECT_EQ(arr.getSize(), 0);
}

TEST(ArrayTest, ResizeCapacity) {
    Array arr(2); // Начальная емкость 2
    arr.pushBack("1");
    arr.pushBack("2");
    // Должен произойти ресайз
    arr.pushBack("3");
    EXPECT_EQ(arr.getSize(), 3);
    EXPECT_EQ(arr.get(2), "3");
}

TEST(ArrayTest, LoadFromNonExistingFiles) {
    Array arr;
    arr.loadFromText("non_existing.txt");
    EXPECT_EQ(arr.getSize(), 0);
    
    arr.loadFromBinary("non_existing.bin");
    EXPECT_EQ(arr.getSize(), 0);
}

TEST(ArrayTest, InsertAtBoundaries) {
    Array arr;
    arr.pushBack("a");
    arr.insertAt(0, "first"); // В начало
    arr.insertAt(2, "last");  // В конец
    EXPECT_EQ(arr.getSize(), 3);
    EXPECT_EQ(arr.get(0), "first");
    EXPECT_EQ(arr.get(2), "last");
}

TEST(ArrayTest, EmptyArrayOperations) {
    Array arr;
    EXPECT_EQ(arr.getSize(), 0);
    EXPECT_THROW(arr.set(0, "x"), std::out_of_range);
    EXPECT_EQ(arr.find("x"), -1);
}

TEST(ArrayTest, PrintMethod) {
    Array arr;
    arr.pushBack("test1");
    arr.pushBack("test2");
    // Просто проверяем что не падает
    arr.print();
}

File: go/array_test.go
package main

import (
	"os"
	"testing"
)

func TestArrayExceptions(t *testing.T) {
	arr := NewArray(10)

	_, err := arr.Get(-1)
	if err == nil {
		t.Error("Expected error for Get(-1)")
	}

	_, err = arr.Get(0)
	if err == nil {
		t.Error("Expected error for Get(0) on empty array")
	}

	err = arr.RemoveAt(0)
	if err == nil {
		t.Error("Expected error for RemoveAt(0) on empty array")
	}

	err = arr.InsertAt(1, "x")
	if err == nil {
		t.Error("Expected error for InsertAt(1, 'x') on empty array")
	}
}

func TestArraySaveLoadTextBinary(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("a")
	arr.PushBack("b")

	err := arr.SaveToText("arr.txt")
	if err != nil {
		t.Fatal(err)
	}

	arr2 := NewArray(10)
	err = arr2.LoadFromText("arr.txt")
	if err != nil {
		t.Fatal(err)
	}

	if arr2.GetSize() != 2 {
		t.Errorf("Expected size 2, got %d", arr2.GetSize())
	}

	val, err := arr2.Get(0)
	if err != nil || val != "a" {
		t.Errorf("Expected 'a', got %v", val)
	}

	err = arr.SaveToBinary("arr.bin")
	if err != nil {
		t.Fatal(err)
	}

	arr3 := NewArray(10)
	err = arr3.LoadFromBinary("arr.bin")
	if err != nil {
		t.Fatal(err)
	}

	if arr3.GetSize() != 2 {
		t.Errorf("Expected size 2, got %d", arr3.GetSize())
	}

	val, err = arr3.Get(1)
	if err != nil || val != "b" {
		t.Errorf("Expected 'b', got %v", val)
	}

	os.Remove("arr.txt")
	os.Remove("arr.bin")
}

func TestArrayBasicOperations(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("1")
	arr.PushBack("3")
	arr.PushFront("0")
	arr.InsertAt(2, "2")

	if arr.GetSize() != 4 {
		t.Errorf("Expected size 4, got %d", arr.GetSize())
	}

	val, _ := arr.Get(0)
	if val != "0" {
		t.Errorf("Expected '0', got %s", val)
	}

	val, _ = arr.Get(1)
	if val != "1" {
		t.Errorf("Expected '1', got %s", val)
	}

	val, _ = arr.Get(2)
	if val != "2" {
		t.Errorf("Expected '2', got %s", val)
	}

	val, _ = arr.Get(3)
	if val != "3" {
		t.Errorf("Expected '3', got %s", val)
	}

	arr.PopFront()
	arr.PopBack()
	if arr.GetSize() != 2 {
		t.Errorf("Expected size 2, got %d", arr.GetSize())
	}
}

func TestArrayFindAndSet(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("a")
	arr.PushBack("b")
	arr.PushBack("c")

	if arr.Find("b") != 1 {
		t.Errorf("Expected index 1 for 'b', got %d", arr.Find("b"))
	}

	if arr.Find("d") != -1 {
		t.Errorf("Expected -1 for 'd', got %d", arr.Find("d"))
	}

	arr.Set(1, "x")
	val, _ := arr.Get(1)
	if val != "x" {
		t.Errorf("Expected 'x', got %s", val)
	}
}

func TestArrayRemoveAtValid(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("a")
	arr.PushBack("b")
	arr.PushBack("c")

	arr.RemoveAt(1)
	if arr.GetSize() != 2 {
		t.Errorf("Expected size 2, got %d", arr.GetSize())
	}

	val, _ := arr.Get(0)
	if val != "a" {
		t.Errorf("Expected 'a', got %s", val)
	}

	val, _ = arr.Get(1)
	if val != "c" {
		t.Errorf("Expected 'c', got %s", val)
	}
}

func TestArrayPopOperationsOnEmpty(t *testing.T) {
	arr := NewArray(10)
	arr.PopBack()
	arr.PopFront()
	if arr.GetSize() != 0 {
		t.Errorf("Expected size 0, got %d", arr.GetSize())
	}
}

func TestArrayResizeCapacity(t *testing.T) {
	arr := NewArray(2)
	arr.PushBack("1")
	arr.PushBack("2")
	arr.PushBack("3")

	if arr.GetSize() != 3 {
		t.Errorf("Expected size 3, got %d", arr.GetSize())
	}

	val, _ := arr.Get(2)
	if val != "3" {
		t.Errorf("Expected '3', got %s", val)
	}
}

func TestArrayLoadFromNonExistingFiles(t *testing.T) {
	arr := NewArray(10)
	err := arr.LoadFromText("non_existing.txt")
	if err == nil {
		t.Error("Expected error for non-existing file")
	}

	err = arr.LoadFromBinary("non_existing.bin")
	if err == nil {
		t.Error("Expected error for non-existing binary file")
	}

	if arr.GetSize() != 0 {
		t.Errorf("Expected size 0, got %d", arr.GetSize())
	}
}

func TestArrayInsertAtBoundaries(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("a")
	arr.InsertAt(0, "first")
	arr.InsertAt(2, "last")

	if arr.GetSize() != 3 {
		t.Errorf("Expected size 3, got %d", arr.GetSize())
	}

	val, _ := arr.Get(0)
	if val != "first" {
		t.Errorf("Expected 'first', got %s", val)
	}

	val, _ = arr.Get(2)
	if val != "last" {
		t.Errorf("Expected 'last', got %s", val)
	}
}

func TestArrayEmptyArrayOperations(t *testing.T) {
	arr := NewArray(10)
	if arr.GetSize() != 0 {
		t.Errorf("Expected size 0, got %d", arr.GetSize())
	}

	err := arr.Set(0, "x")
	if err == nil {
		t.Error("Expected error for Set on empty array")
	}

	if arr.Find("x") != -1 {
		t.Errorf("Expected -1 for 'x' in empty array, got %d", arr.Find("x"))
	}
}

func TestArrayPrintMethod(t *testing.T) {
	arr := NewArray(10)
	arr.PushBack("test1")
	arr.PushBack("test2")
	arr.Print()
}
